### DISCUSSION TITLE
Road to 0.68 - part 2

### DISCUSSION THREAD

**Discussion by kelset** [reactwg/react-native-releases#17] [Releases]
# 0.68.0 is RELEASED! Conversation continues here: https://github.com/reactwg/react-native-releases/discussions/19

---

The branch cut has happened.

## Notice

* [Current release candidate: 0.68.0-rc.4][current-release]
* Have an issue with current release candidate? [File an issue][issue-form] and we will triage.
* Have a pick request for this release? Does it fall under our [pick request qualifications][faq-wiki]? If so please create a PR against the release branch and comment with the PR link
* If you are release testing. Copy and fill [Release report form](https://github.com/facebook/react-native/wiki/Release-Tester-Matrix)

#### Highlighted Changes in this release
 
- Checkout this [Changelog PR][changelog-pr]
- RN0.68 will use AGP 7.0.x, **not** 7.1. See [this comment](https://github.com/reactwg/react-native-releases/discussions/13#discussioncomment-2221683) for more details.

#### Checklist
- [x] [Changelog PR][changelog-pr]
- [x] Start a Google doc of blog post for release and invite contributors of release highlights to expand
- [x] cut new docs version on react-native-website https://github.com/facebook/react-native-website/pull/3033
- [x] Follow up on [release dependencies][dependencies-wiki]
#### When ready to publish stable
- [ ] Ship changelog https://github.com/facebook/react-native/pull/32993
- [ ] Ship blog post https://github.com/facebook/react-native-website/pull/3041
- [ ] Notify `react-native-website` to ship new version https://github.com/facebook/react-native-website/pull/3033

#### Retrospective Topics

* *add me*

#### Tracking issues for 0.68.0-rc.5
* *add me*

#### Picks for 0.68.0-rc.5
* *add me*

#### Picks for 0.68.0-rc.4
* https://github.com/facebook/react-native/commit/b5343a6b0dd07c1b4ef9dac549df67a4d68ebd1e

[changelog-pr]: https://github.com/facebook/react-native/labels/%F0%9F%93%9D%20Changelog
[current-release]: https://github.com/facebook/react-native/releases/tag/v0.68.0-rc.4
[changelog-wiki]: https://github.com/facebook/react-native/wiki/Release-Changelog
[dependencies-wiki]: https://github.com/facebook/react-native/wiki/Release-Dependencies
[faq-wiki]: https://github.com/facebook/react-native/wiki/Release-FAQ
[issue-form]: https://github.com/facebook/react-native/issues/new?assignees=&labels=Needs%3A+Triage+%3Amag%3A%2Cpre-release&template=release_blocker_form.yml
[releases]: https://github.com/facebook/react-native/releases
[release-process-wiki]: https://github.com/facebook/react-native/wiki/Release-Process
[upgrade-helper-wiki]: https://github.com/facebook/react-native/wiki/Updating-upgrade-helper

---

For previous RCs conversation, please refer to [the previous discussion](https://github.com/reactwg/react-native-releases/discussions/13).

**Comment by kelset** [comment]
should we pick the bumps to minimum iOS version?
* https://github.com/facebook/react-native/commit/982ca30de079d7e80bd0b50365d58b9048fb628f
* https://github.com/facebook/react-native/commit/c71e6efbcd2b95faee327d9763d321488120bc5e

**Reply by danilobuerger** [reply]
I don't see a real advantage in picking them for 0.68 as this release should just work fine with 11.0 - 12.3, so why limit devs? Devs still supporting 11.0 would then be stuck with 0.67 instead of 0.68.

**Reply by kelset** [reply]
fine by me, that's why I was asking if we should ;)
Given that iOS 15 is current latest, I assumed not many folks would still need 11 around

**Comment by fortmarek** [comment]
Considering there have been some changes in third-party podspecs (such as [here](https://github.com/facebook/react-native/commit/982ca30de079d7e80bd0b50365d58b9048fb628f)), it might be a good idea to pull in this change:
https://github.com/facebook/react-native/commit/b2517c3bdccc3f9d935f4ee06f959d6ce8f27bbe

The version of the third-party local podspecs have not changed afaict though, in which case the commit would not make a difference when bumping from `0.67.x`. It would still update those podspecs automatically if you upgraded from previous React Native versions.

Tangentially, should the version be bumped when the podspec itself is changed or is that not related?

**Reply by mikehardy** [reply]
:wave: from the developing world, and also from a planet struggling with so many humans. When fundamental frameworks bump minimum device requirements it pushes a bunch more old devices - that are the only devices many people have (that "7%" missing in the quote "93% of devices are on ios14") - closer to the landfill. I urge everyone involved in any decision like that to consider carefully if an n-1 policy is important for fundamental technical reasons or if it's just more convenient.

**Reply by fortmarek** [reply]
Hey @mikehardy 👋  
To clarify, I am not advocating for making the bump of iOS version. The commit linked actually is not even cherry-picked (my mistake). The rest of the points still stand, though.

**Reply by fortmarek** [reply]
Given we are close to actually making a release, this commit is not crucial enough to be included in the current release, so I'd drop it and release it with 0.69 instead.

**Comment by Megatron4537** [comment]
Tested commits `run/android/0.68.0-rc3` and everything was good! 

I am curious if this take took a long time for you? for me it would take ~5-10 minutes every gradle clean and new builds
```
> Task :ReactAndroid:downloadLibevent
Download https://github.com/libevent/libevent/releases/download/release-2.1.12-stable/libevent-2.1.12-stable.tar.gz
```

**Reply by cortinico** [reply]
> for me it would take ~5-10 minutes every gradle clean and new builds

That's expected, especially if you do a full clean build and you're on the New Architecture. All the native build needs to be re-executed.

**Reply by danilobuerger** [reply]
For me it was a mismatch between the jdks and my cpus arch. Once I installed an arm64 jdk on my m1 mac, the build time was reduced in half.

**Reply by alpha0010** [reply]
Is it possible to use `ccache`  in the build system?  It makes compilation of C/C++ code essential instant (well, linking objects still takes time) on clean rebuilds.

I have been using it for a large native module, with cmake build system.  15 minute build -> 30 seconds.

Autodetect `ccache`, and use if found:
```cmake
find_program(CCACHE_FOUND ccache)
if(CCACHE_FOUND)
    set_property(TARGET my_library PROPERTY CXX_COMPILER_LAUNCHER ccache)
endif(CCACHE_FOUND)
```

**Reply by mikehardy** [reply]
I'm not sure if that's something to enable by default in the framework, note that brew on purpose will not use it even if installed, you have to manually force it's use for brew and then I think they won't take bug reports. So apparently there is a support cost. (reference, check the note: https://formulae.brew.sh/formula/ccache)

But I personally use it for the iOS part of the compilation and it is quite the accelerator, yes.

That argues for something that *allows* it's use but does not use it by default. Similar to this to *allow* it for iOS compilation but it's certainly not the default https://github.com/invertase/react-native-firebase-authentication-example/blob/68bd37f8c219f892655b40c73e9818370f368b82/template/ios/Podfile#L39-L43

**Reply by leotm** [reply]
> For me it was a mismatch between the jdks and my cpus arch. Once I installed an arm64 jdk on my m1 mac, the build time was reduced in half.

similar here after upgrading _Android Studio Preview Canary_ via the [site](https://developer.android.com/studio/preview) or [brew](https://formulae.brew.sh/cask/android-studio) ([cask](https://github.com/Homebrew/homebrew-cask-versions/blob/master/Casks/android-studio-preview-canary.rb))

think my JDK's resetting to _AdoptOpenJDK 11_ each time, then ok again after [reverting](https://github.com/leotm/react-native-template-typescript/wiki/JDK#eclipse-temurin-11) back to (Adoptium) Eclipse Temurin 11

**Reply by mikehardy** [reply]
> > For me it was a mismatch between the jdks and my cpus arch. Once I installed an arm64 jdk on my m1 mac, the build time was reduced in half.
> 
> think my JDK's resetting to _AdoptOpenJDK 11_ each time, then ok again after [reverting](https://github.com/leotm/react-native-template-typescript/wiki/JDK#eclipse-temurin-11) back to (Adoptium) Eclipse Temurin 11

@leotm 🤔 but adoptium temurin 11 is only x86_64 right now (I just checked). There are plans for arm64 backport to release soon for Temurin11 but right now only arm64 JDK11 is Azul Zuzu.  If you are reverting to Temurin 11 right now, you are not getting the native-arm64 speed boost that @danilobuerger was describing

**Reply by leotm** [reply]
gotcha rip my m1 android build bens this whole time 😅 been sticking w [docs (next)](https://reactnative.dev/docs/next/environment-setup) [(this part)](https://user-images.githubusercontent.com/1881059/159939610-c6ed60f4-9e6c-481a-800d-d943b24fddd1.PNG)
 
i'll raise that M1/arm64 JDK note @mikehardy in https://github.com/reactwg/react-native-new-architecture/discussions/7 if i'm understanding right (and strictly note perf metrics from now on)

edit: <s>actually that doesn't look like the right place, i'll find the repo of the docs and PR a note/suggestion</s>
[Edit this page](https://github.com/facebook/react-native-website/blob/main/docs/getting-started.md) doesn't seem like the right place either 🤔

and figure if it's pos stopping JDK reset per upgrade

---

i remember now in [github.com/actions/setup-java#basic](https://github.com/actions/setup-java#basic) Zulu OpenJDK 2nd option

not Zuzu (; thx again

**Reply by cortinico** [reply]
We're going to have an official guide about this specific issue here:
- https://github.com/facebook/react-native-website/pull/3027

**Comment by kelset** [comment]
we need this: https://github.com/facebook/react-native/pull/33499 - commit https://github.com/facebook/react-native/commit/b5343a6b0dd07c1b4ef9dac549df67a4d68ebd1e

**Comment by tido64** [comment]
@cortinico: `react-native-gradle-plugin` is not requiring the same version of `react-native-codegen` as core does. Is this going to be an issue?
```
% yarn why react-native-codegen
├─ react-native-gradle-plugin@npm:0.0.5
│  └─ react-native-codegen@npm:0.0.14 (via npm:*)
│
└─ react-native@npm:0.68.0-rc.4 [6dfc2]
   └─ react-native-codegen@npm:0.0.13 (via npm:^0.0.13)
```

**Reply by cortinico** [reply]
~Not really, as `react-native-gradle-plugin` is not having any js code, and is not consuming any dependency. That dependency could probably be removed entirely~

EDIT: Actually, It might. I'm looking into it as we might have to remove that dependency (react-native-gradle-plugin -> react-native-codegen) entirely for 0.68.0.

**Reply by cortinico** [reply]
Just checked and nope, it's not a problem. 

You're going to have `react-native-codegen@npm:0.0.14` inside `node_modules/react-native-gradle-plugin/node_modules/react-native-codegen/` which is, I agree, not great.
However, as `react-native-gradle-plugin` contains no `js` code, that dependency is essentially ignored. 

To mitigate this scenarios, I will replace the `"*"` from https://github.com/facebook/react-native/blob/d2e8e7d58e680e0bb3b4da1f820dd4dd840639f5/packages/react-native-gradle-plugin/package.json#L27 with the latest released version of `"react-native-codegen"` so this won't happen again in the future.

**Comment by douglowder** [comment]
@kelset should https://github.com/facebook/react-native/pull/33500 (https://github.com/facebook/react-native/commit/58a2eb7f37c2dc27ad3575618778ad5b23599b27) be pulled in?

**Reply by kelset** [reply]
not sure - it's a bit unclear to me how widespread the problem it tackles is. Maybe it's worth considering for 0.68.1 instead? I'll defer to @ShikaSD / @cortinico since it's Android specific (and seems related to new arch?)

**Reply by ShikaSD** [reply]
It is needed for Reanimated to work with the new arch, I believe we can skip it for now

**Reply by douglowder** [reply]
Thanks @kelset @ShikaSD ... I can cherry-pick to react-native-tvos if needed

**Reply by cortinico** [reply]
> I believe we can skip it for now

Agree. I think we can point release, if Reanimated fully migrates to new arch.
